Doc/reference/expressions.rst

```diff
@@ -27,11 +27,12 @@ Arithmetic conversions
 
 .. index:: pair: arithmetic; conversion
 
-When a description of an arithmetic operator below uses the phrase "the numeric
+When a description of an arithmetic operator below uses the phrase "the real numeric
 arguments are converted to a common type", this means that the operator
 implementation for built-in types works as follows:
 
-* If either argument is a complex number, the other is converted to complex;
+* If only one argument is a complex number, the other is converted to a
```

Comment:
According to this algorithm, if both arguments are complex, we will fall into "otherwise, both must be integers and no conversion is necessary."

Add "If both arguments are complex numbers, no conversion is necessary" as the first item. Or rewrite is something like:

* If both arguments are a complex number, no conversion is necessary;
* otherwise, if either argument is a complex or floating-point number, the other is
  converted to floating point;
* otherwise, both must be integers and no conversion is necessary.

This will need other rewriting after adding imaginary numbers.

On other hand, the only conversion is from integer to floating point if other argument is complex or floating point. "converted to a common type" is now a lie. We should use different word to describe such conversion.

* If one argument is a complex or floating-point number and the other is an integer, the latter one is
  converted to floating point.

Reply:
> This will need other rewriting after adding imaginary numbers.

In principle, there is a possibility to avoid of a new type.  What if e.g. ``complex(imag=x)`` will make special instances of ``complex`` type, that will serve as imaginary numbers (i.e. no real part in arithmetic rules)?

> "converted to a common type" is now a lie

It's not.  Original version "the numeric arguments are converted to a common type" was replaced with "the numeric arguments are converted to a common real type".

Maybe it could be better, but I think current version does make sense: after conversion, components of the complex number and other operand - have common type.

Reply:
I think we can avoid changes in ``stdtypes.rst`` and ``expressions.rst``.  Used previously wording "type is widened" or "converted to a common type" is vague enough.

But how exactly it's "converted"?  E.g. for a float ``x``, is this ``complex(x)`` or ``complex(x, 0.0)`` or ``complex(real=x)``?  Now these forms of the constructor are equal, but we could change that.   What if the last one (hardly popular) will means "a complex number on the real axis of the complex plane"?  Currently we don't have such a concept: ``complex(x, 0.0)`` means actually "above real line", just like ``complex(x, -0.0)`` - below.

In a similar way we might treat the imaginary axis: ``complex(imag=y)``.  No new type.

With this, following identities will hold:
```
    complex(x, y) == complex(real=x, imag=y)
                  == complex(real=x) + complex(imag=y)
```

The price:
* new meaning for ``complex(real=x)`` (but such numbers will be visible for user only if explicitly constructed)
* it's repr, probably, should be verbose: ``f"complex(real={x})"``.

if with the imaginary axis as well:
* new meaning for ``complex(imag=y)`` (the repr will be ``f"{y}j"``)
* ``repr(complex(±0.0, y))`` will be changed to ``f"±0.0{y:+}j"`` (now with ``-0.0`` we have funny form with signed integer zero and with 0.0 - real part is omitted)
* ``complex(real=x)`` forms will appear in computations (e.g. ``1j*1j``), unless we consider to return just ``-y*z`` (float type) for ``complex(imag=y)*complex(imag=z)``.  The later does make sense, as ``complex(real=x)`` set is isomorphic to the real line (usual floats).

Reply:
@picnixz, thank you for review.  I hope I addressed your points.

Probably, the documentation part is most controversial now.  I would appreciate your opinion on above approach.

Reply:
Ah sorry, I forgot to reply on this one.

Reply:
> What if e.g. complex(imag=x) will make special instances of complex type, that will serve as imaginary numbers (i.e. no real part in arithmetic rules)?

I think we can either postpone this issue or take what people expect complex numbers to be in general, not what we could make them look like. To be a "true" real (and not just above/below), I think we should rather introduce a new type, or perhaps an attribute or something else. I'm pretty sure that we would break scripts that don't make this distinction.

>  Currently we don't have such a concept: complex(x, 0.0) means actually "above real line", just like complex(x, -0.0) - below.

I think it's fine for now. Dealing with purely real and purely imaginary complex numbers would probably require more work than just changing this specific place of the code and I think it's the responsibility to the numerical library that will need them (e.g., SAGE or NumPy). I know that it's not really nice saying "that's not our problem for now", but unless there is a real need for having the possibility to represent purely imaginary and purely real complexes, we should be as simple as possible.

Reply:
In other words: I'm fine with what the formulation you suggested in this PR and I'm fine with not explaining the concept of purely imaginary / purely real complex numbers in detail.

Reply:
> To be a "true" real (and not just above/below), I think we should rather introduce a new type, or perhaps an attribute or something else.

We have such type, it's the float type.  The question is: how to convert one to the complex type?

> Dealing with purely real and purely imaginary complex numbers would probably require more work

Not at all. Current PR uses the float type to represent "purely real" numbers. But that requires some documentation changes to document conversions in binary ops. If we map floats to complex(real=x) - I think that most from these changes could be reverted.